File: fix_order_model.dart
import 'dart:io';

void main() {
  final file = File('lib/models/order_model.dart');
  var content = file.readAsStringSync();
  
  // 1. Add fields to class
  content = content.replaceFirst(
    'final DateTime createdAt;\n  final DateTime? completedAt;\n  final String deliveryNote;',
    'final DateTime createdAt;\n  final DateTime? completedAt;\n  final DateTime? startedAt;\n  final String deliveryNote;\n  final String deliveryLink;\n  final String revisionNote;'
  );

  // 2. Add fields to constructor
  content = content.replaceFirst(
    'this.completedAt,\n    this.deliveryNote = \'\',',
    'this.completedAt,\n    this.startedAt,\n    this.deliveryNote = \'\',\n    this.deliveryLink = \'\',\n    this.revisionNote = \'\','
  );

  // 3. Add fields to fromFirestore
  content = content.replaceFirst(
    'completedAt: (d[\'completedAt\'] as Timestamp?)?.toDate(),\n      deliveryNote: d[\'deliveryNote\'] ?? \'\',',
    'completedAt: (d[\'completedAt\'] as Timestamp?)?.toDate(),\n      startedAt: (d[\'startedAt\'] as Timestamp?)?.toDate(),\n      deliveryNote: d[\'deliveryNote\'] ?? \'\',\n      deliveryLink: d[\'deliveryLink\'] ?? \'\',\n      revisionNote: d[\'revisionNote\'] ?? \'\','
  );

  // 4. Add fields to toMap
  content = content.replaceFirst(
    '\'createdAt\': FieldValue.serverTimestamp(),\n      \'deliveryNote\': deliveryNote,',
    '\'createdAt\': FieldValue.serverTimestamp(),\n      \'deliveryNote\': deliveryNote,\n      \'deliveryLink\': deliveryLink,\n      \'revisionNote\': revisionNote,'
  );
  content = content.replaceFirst(
    'if (completedAt != null) \'completedAt\': Timestamp.fromDate(completedAt!),',
    'if (completedAt != null) \'completedAt\': Timestamp.fromDate(completedAt!),\n      if (startedAt != null) \'startedAt\': Timestamp.fromDate(startedAt!),'
  );

  file.writeAsStringSync(content);
  print('Updated order_model.dart');
}

File: fix_order_model.py
with open('lib/models/order_model.dart', 'r') as f:
    c = f.read()

import re

# Helper functions for robust parsing
old_factory = "  factory OrderModel.fromFirestore(DocumentSnapshot doc) {"
new_factory = """  static double? _parseDouble(dynamic value) {
    if (value == null) return null;
    if (value is num) return value.toDouble();
    if (value is String) return double.tryParse(value);
    return null;
  }

  static int? _parseInt(dynamic value) {
    if (value == null) return null;
    if (value is num) return value.toInt();
    if (value is String) return int.tryParse(value);
    return null;
  }
  
  static DateTime? _parseDate(dynamic value) {
    if (value == null) return null;
    if (value is Timestamp) return value.toDate();
    if (value is int) return DateTime.fromMillisecondsSinceEpoch(value);
    if (value is String) return DateTime.tryParse(value);
    return null;
  }

  factory OrderModel.fromFirestore(DocumentSnapshot doc) {"""

c = c.replace(old_factory, new_factory)

# Replace fields
c = c.replace("price: (d['price'] as num?)?.toDouble() ?? 0,", "price: _parseDouble(d['price']) ?? 0,")
c = c.replace("basePrice: (d['basePrice'] as num?)?.toDouble() ?? 0,", "basePrice: _parseDouble(d['basePrice']) ?? 0,")
c = c.replace("discountAmount: (d['discountAmount'] as num?)?.toDouble() ?? 0,", "discountAmount: _parseDouble(d['discountAmount']) ?? 0,")
c = c.replace("deliveryDays: (d['deliveryDays'] as num?)?.toInt() ?? 3,", "deliveryDays: _parseInt(d['deliveryDays']) ?? 3,")

c = c.replace("ratingCommunication: (d['ratingCommunication'] as num?)?.toDouble(),", "ratingCommunication: _parseDouble(d['ratingCommunication']),")
c = c.replace("ratingQuality: (d['ratingQuality'] as num?)?.toDouble(),", "ratingQuality: _parseDouble(d['ratingQuality']),")
c = c.replace("ratingDescribed: (d['ratingDescribed'] as num?)?.toDouble(),", "ratingDescribed: _parseDouble(d['ratingDescribed']),")
c = c.replace("overallRating: (d['overallRating'] as num?)?.toDouble(),", "overallRating: _parseDouble(d['overallRating']),")

c = c.replace("createdAt: (d['createdAt'] as Timestamp?)?.toDate() ?? DateTime.now(),", "createdAt: _parseDate(d['createdAt']) ?? DateTime.now(),")
c = c.replace("completedAt: (d['completedAt'] as Timestamp?)?.toDate(),", "completedAt: _parseDate(d['completedAt']),")
c = c.replace("startedAt: (d['startedAt'] as Timestamp?)?.toDate(),", "startedAt: _parseDate(d['startedAt']),")
c = c.replace("deliveredAt: (d['deliveredAt'] as Timestamp?)?.toDate(),", "deliveredAt: _parseDate(d['deliveredAt']),")

with open('lib/models/order_model.dart', 'w') as f:
    f.write(c)
